### PR TITLE
[NAVAND-935] do not interrupt reroute when alternative or refreshed routes are set

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

This PR adjust the `MapboxNavigation` to avoid interrupting the reroute controller when alternative or refresh are set. These calls do not fully reset the primary route, which means that we'll remain in an `off-route` state but the `OffRouteObserver` won't fire again and we'll be stuck with a route user deviated from without any task which tries to fetch a new instance.

Additionally, I added a commit that retries the reroute if Nav Native rejects the newly generated route to prevent user getting stuck. I might extract it out to a separate PR though.

Cutting as a draft with tests to follow.

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt [hunk 1]
```diff
@@ -151,8 +151,8 @@ internal class MapboxRerouteController @VisibleForTesting constructor(
                 val origin = relevantAlternative.routerOrigin.mapToSdkRouteOrigin()
 
                 state = RerouteState.RouteFetched(origin)
-                callback.onNewRoutes(newList, origin)
                 state = RerouteState.Idle
+                callback.onNewRoutes(newList, origin)
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:155]: Why is this needed?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 2]
```diff
@@ -1967,7 +1979,15 @@ class MapboxNavigation @VisibleForTesting internal constructor(
...
                 routes,
                 SetRoutes.Reroute
-            )
+            ) {
+                it.onError {
+                    logE(LOG_CATEGORY) {
+                        "Failed to set route generated by the reroute controller, " +
+                            "reason: ${it.message}. Retrying."
+                    }
+                    reroute()
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1988]: Do we expect to get a new (better) route here? Why would it change?
I'm worried that we will be forever rerouting in a loop if something goes wrong. Can we set the maximum number of retries?

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1988]: > Do we expect to get a new (better) route here? Why would it change?

This is in case Nav Native fails to accept the route, so possibly a somewhat malformed or outdated response from the backend.

I'll cut a separate PR with this.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt (hunk 1 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:155]: This is needed to make sure that we can request a new reroute from an existing reroute callback. I'll also move this to a separate PR.

**Comment by LukasPaczos** [comment]
This is ready for review.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 3]
```diff
@@ -281,19 +273,195 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+            relaxedExpectedCoordinates = false
+        )
+        val rerouteRequestHandler = MockDirectionsRequestHandler(
+            profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+            jsonResponse = mockRouteReroute.routeResponseJson,
+            expectedCoordinates = listOf(
+                offRouteLocation,
+                mockRoute.routeWaypoints.last()
+            ),
+            relaxedExpectedCoordinates = true
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:310]: I've always wondered why we bother passing the coordinates if they are relaxed...

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 4]
```diff
@@ -281,19 +273,195 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+        )
+        val responseModifier = DelayedResponseModifier(4000) { it }
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:312]: NIT: you can drop the last argument: it will return the original value by default.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 5]
```diff
@@ -281,19 +273,195 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(originalRoutes.first()))
+
+        mapboxNavigation.routeProgressUpdates().filter {
+            it.currentState == RouteProgressState.TRACKING
+        }.first()
+        mockLocationReplayerRule.stopAndClearEvents()
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 120)
+        // wait for OFF_ROUTE
+        mapboxNavigation.offRouteUpdates().filter { it }.first()
+        delay(2000)
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:347]: Why?
I hate delays because they may introduce flakiness. Although I understand that sometimes they are the only solution. If it's the case here, can you add a comment stating why it's required?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 6]
```diff
@@ -281,19 +273,195 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+        mockWebServerRule.requestHandlers.add(rerouteRequestHandler)
+        val refreshHandler = MockDirectionsRefreshHandler(
+            testUuid = "jpCHHUC26qFOwISCNLjar2xmTfI6Dxd0qCHOoqwt_1VAlESNvsr7Zg==",
+            readRawFileText(
+                activity,
+                R.raw.route_response_dc_short_with_alternative_refresh_route_0
+            ),
+            acceptedGeometryIndex = 0
+        )
+        mockWebServerRule.requestHandlers.add(refreshHandler)
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:415]: There is a lot of duplicated code (in this test and the test above) and it doesn't really say anything about the test cases. I mean the setup of refresh handlers, etc. Can we extract it to a method so that it doesn't take up a lot of lines in the test itself?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 7]
```diff
@@ -281,19 +273,195 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+        mapboxNavigation.routeProgressUpdates().filter {
+            it.currentState == RouteProgressState.TRACKING
+        }.first()
+        mockLocationReplayerRule.stopAndClearEvents()
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 120)
+        // wait for OFF_ROUTE
+        mapboxNavigation.offRouteUpdates().filter { it }.first()
+        // wait for refresh update
+        mapboxNavigation.routesUpdates().first {
+            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:449]: Aren't you afraid that the reroute will finish before you register the observer? Won't it break anything? I mean you have:
```
mapboxNavigation.routesUpdates().first {
            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH
}
val routesUpdate = mapboxNavigation.routesUpdates().first {
            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
}
```
I'm worried one of the line might get stuck.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/routes/RoutesProvider.kt [hunk 8]
```diff
@@ -80,8 +80,30 @@ object RoutesProvider {
     fun dc_short_with_alternative(context: Context): MockRoute {
         val jsonResponse = readRawFileText(context, R.raw.route_response_dc_short_with_alternative)
         val coordinates = listOf(
-            Point.fromLngLat(-77.033991, 38.891121),
-            Point.fromLngLat(-77.030023, 38.895433)
+            Point.fromLngLat(-77.033625, 38.891164),
```
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/routes/RoutesProvider.kt:83]: Won't it break old tests? We might use the hardcoded coordinates somewhere...

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 3 above)
Comment (LukasPaczos) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:310]: Otherwise we'd have no way to distinguish the requests if there are multiple in one test. But I agree that this is not ideal.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 5 above)
Comment (LukasPaczos) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:347]: I wanted to set the alternatives async to reflect the exact issue I'm testing against, and which is much more likely in real world. Alternatives wouldn't be set synchronously at the same time as off-route happened.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 7 above)
Comment (LukasPaczos) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:449]: The refresh delay in the test is 3 seconds, the reroute 4 seconds, so I'm expecting the refresh to happen before the reroute. If that wasn't the case, this code would indeed get stuck and time out but that's good as it would give us the visibility into the fact the test is not testing what it's meant to be testing anymore.

Unfortunately, I have only 5 seconds to work with because that's the default timeout in Nav Native. We can refactor that in the future when we expose the granular request timeouts option

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/routes/RoutesProvider.kt (hunk 8 above)
Comment (LukasPaczos) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/routes/RoutesProvider.kt:83]: This route is used only in reroute and preview tests. The `dc_very_short` variant is much more popular but that's not the one I updated.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 3 above)
Comment (LukasPaczos) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:310]: Well, looks like `relaxedExpectedCoordinates` is more like `ignoreExpectedCoordinates`, so I agree with what you're saying :)

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 9]
```diff
@@ -1015,8 +1016,21 @@ class MapboxNavigation @VisibleForTesting internal constructor(
...
+        when (setRoutesInfo) {
+            SetRoutes.CleanUp,
+            is SetRoutes.NewRoutes,
+            SetRoutes.Reroute -> {
+                rerouteController?.interrupt()
+            }
+            is SetRoutes.RefreshRoutes,
+            is SetRoutes.Alternatives -> {
+                // do not interrupt reroute when primary route has not changed
+            }
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1031]: Was thinking about this for a little, It's worth adding definitions to `SetRoutes` because logic depends on the definition. Thinking about scenarios where someone should be calling `NewRoutes` but they decide to call `SetRoutes.Alternatives`.

```

    /** 
     * Triggered when the alternative routes change and the primary route has remained the same.
     */
    internal data class Alternatives(
        val legIndex: Int,
    ) : SetRoutes()
```

**Review by kmadsen — APPROVED** [review]
LGTM

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 5 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:321]: Oh, so this delay guards nothing, it's just imitating real world? Phew, that's a relief. Could you add a comment stating that?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 7 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:402]: Ah, now I see. Didn't pay attention to the intervals and delays. Just for clarity it would be nice to add something like:
```
// refreshInterval has to be less than the reroute delay as we expect the refresh to happen before the reroute
val refreshInterval = 3000
// reroute delay has to be less than 5 seconds as it's the default timeout in NN
val rerouteDelay = 4000
```

**Review by dzinad — APPROVED** [review]
